### ISSUE TITLE
unused variable: warn about collections that are pushed to but not used afterwards

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#49151]
````rust
fn main() {
        let mut var: Vec<i64> = Vec::new();
        var.push(1);
}
````
I would expect a warning here that ````var```` is unused.
I get that it is "used" in the sense that we push to it, however the *resulting* vector is still not done anything with.

Some kind of dead-store warning about the collection being unused/never queried after receiving an item might be helpful.

**Comment by matthiaskrgr** [comment]
Just for the record, if I rewrite the code as follows
````rust
fn main() {
    let var: Vec<i64> = vec![1];
}
````
I get a warning
````
warning: unused variable: `var`
 --> src/main.rs:2:9
  |
2 |     let var: Vec<i64> = vec![1];
  |         ^^^ help: consider using `_var` instead
  |
  = note: #[warn(unused_variables)] on by default
````

**Comment by durka** [comment]
I don't see any way to do this in a generalizable way.

```rust
let mut var = Missile::new();
var.launch();
```

Yeah, `var` is not consumed, but...

**Comment by matthiaskrgr** [comment]
Hmm, I mostly had the ````std::collections```` in mind, which we might be able to savely reason about.

**Comment by abonander** [comment]
What about an attribute for methods like `#[does_not_use]` that tells the unused-variables lint that those method calls shouldn't mark a value used. That could cause false-positives like here:

    fn add_something(vals: &mut Vec<i32>) {
        vals.push(1);
    }
  
    fn main() {
        let mut vals = Vec::new();
        add_something(vals);
        println!("{:?}", vals);
    }

But we could have an exception for mutable references (though I'm not sure if that's easy to deduce during the linting pass)

**Comment by oli-obk** [comment]
This feels very much like something that clippy can solve nicely for now by blacklisting certain types and methods
